fabble上のプロジェクトを一覧表示する

noriyotcp opened the pull request:
Closes #21 

## Before
空のクエリを投げると、このように全てのプロジェクトの一覧が表示される
![fabble-all-projects](https://user-images.githubusercontent.com/5820754/43320346-268a9050-91e3-11e8-8b9c-c3fdff31ffd5.png)


## After
空のクエリを付加したリンクをヘッダー部分に設置した

![fabble-header-with-all-projects-link](https://user-images.githubusercontent.com/5820754/43320148-6db41c2c-91e2-11e8-8267-ac0fca686bcd.png)

@takeyuweb このように検索ボックスの横に設置してみました。細かな見た目などの調整ありましたらご指摘ください

-----

### その他

空で検索した場合のurlは、元々このようになっていました

`http://localhost:3000/search?utf8=%E2%9C%93&q=`

```sh
Started GET "/search?utf8=%E2%9C%93&q=" for ::1 at 2018-07-27 19:30:00 +0900
Processing by GlobalProjectsController#index as HTML
  Parameters: {"utf8"=>"✓", "q"=>""}
```

`utf8` は多分これでしょうか
https://qiita.com/tyamagu2/items/fbc44f7b41a1697ca5b7

今回はフォームでなく、リンク置いただけなのでurlは `http://localhost:3000/search?q=` こうなっております

itkrt2y commented on app/views/layouts/_header.html.slim:
```diff
@@ -10,6 +10,9 @@ section#layouts-header
           / TODO: add clear btn
 
+      #all-projects-link
+        == link_to "All Projects", search_path(q: "")
```
`== link_to` => `= link_to`

itkrt2y replied:
そもそも論としては、 `global_projects#index` を `projects#index` に統合し検索クエリなしで動くようにして、 `projects#index` にリンクすべき

noriyotcp commented:
@takeyuweb  @itkrt2y 

すみません!!! すっかり放置していたのですが、今頃急に、空クエリを投げた場合のテストをしていなかったのを思い出したので追加しました

もともと @takeyuweb さんにリクエストしていたのは、リンクの位置や見た目などの確認をお願いしたかったからです
あくまで仮デザインとして、とりあえず検索フォームの横、スタイルも最低限の調整だけにしています

itkrt2y commented on app/views/layouts/_header.html.slim:
```diff
@@ -10,6 +10,9 @@ section#layouts-header
           / TODO: add clear btn
 
+      #all-projects-link
+        = link_to "All Projects", search_path(q: "")
```
`(q: "")` を消す

noriyotcp replied:
それを消すと、`GlobalProjectsController#index` の`else` 以降になってしまいますね
つまり通常のトップページと同じ表示になります

https://github.com/takeyuwebinc/gitfab2/blob/master/app/controllers/global_projects_controller.rb#L18

itkrt2y replied:
そのURL設計、直したいって思いません？

itkrt2y replied:
現状、root（ `/` ）と検索先（ `/search` ）がglobal_projects#indexに同居している。

https://github.com/takeyuwebinc/gitfab2/blob/ac5224f1c8e252b672293757129a6c303e092a0b/config/routes.rb#L16
https://github.com/takeyuwebinc/gitfab2/blob/ac5224f1c8e252b672293757129a6c303e092a0b/config/routes.rb#L27
https://github.com/takeyuwebinc/gitfab2/blob/ac5224f1c8e252b672293757129a6c303e092a0b/config/routes.rb#L77

これはつまり別のURLをわざわざ一つのアクションで受けられるようにしているのにアクション内ではURLごとに固定の処理をしているということで、別のものを一つにまとめてまた元どおりに分解し直していて無駄。

だったら最初から別に定義しておけばいい。

noriyotcp replied:
確かに(q: "") を消した場合に、遷移先が`/search` になるにも関わらず、通常のトップページと同じというのは変ですね

別に定義するべきですね

takeyuweb commented:
@noriyotcp 見た目について、テキストボックスと近すぎて検索ボタンと見間違えたので、そのスペースの中間あたりにできますかね

noriyotcp commented:
WIP: 
まず検索クエリなしの`GlobalProjects#index`と`search` に切り分けたが 2977578

板倉さんが旧`Projects#index`を `Owner#show` へ移動してくれたので https://github.com/takeyuwebinc/gitfab2/pull/36#pullrequestreview-143853156 の統合案が可能になった fa6bbbf

通常のトップページは`Projects#index` に紐づいたが、`search` はまだ`GlobalProjects` 側に残っている

必要なパーシャルをガッと`projects` のほうへ移してしまったので、クリーンアップが必要。パーシャル多すぎる

`global_projects` は消してしまいたいが、`layouts/project.html.slim` が曲者で、`Projects#index`にアクセスした際に `if @project.persisted?` で`@project` がなくて死んでしまう

なので`Projects#index`内で `render layout: 'global_projects'` している状態

noriyotcp commented:
@takeyuweb リンクの位置について、このくらいでどうでしょうか？

![2018-08-16 12 31 05](https://user-images.githubusercontent.com/5820754/44185754-5af10280-a150-11e8-9439-88f15d1be931.png)

noriyotcp commented:
WIP:
`global_projects` を削除した
旧`GlobalProjects#index`, `#search` は`Projects#index`, `#search` へ移行させた

それに伴いスタイルも旧`projects/index.scss` から `owners/show.scss` へ、セレクタも`#projects-index` -> `#owners-show` にした

旧`global_projects/index.scss` -> `projects/index.scss` へと移行

`projects/` 配下のパーシャルがとにかく多いので、なんとかする

noriyotcp committed:
Move style from projects/index to owners/show

旧Projects#indexがOwners#showに移行したことに伴い、スタイルも移行させた

Rename #projects-index to #owners-show

noriyotcp committed:
Remove global_projects

旧GlobalProjects#index, #searchはProjects#index, #searchへ移行済み

noriyotcp commented on app/assets/stylesheets/projects/comment.scss:
```diff
@@ -109,6 +109,7 @@
           font-family: $common-novelsans-light;
           font-size: 12px;
           line-height: 19px;
+          word-wrap: break-word;
```
`CardComment` で文字数が多い時にはみ出していたのを修正

![2018-08-20 18 55 05](https://user-images.githubusercontent.com/5820754/44335008-0f22be00-a4ae-11e8-8f4b-c1224fa42573.png)

itkrt2y commented:
PRの本題と関係のないコミットは別にPRを作ってもらった方が読みやすいですしマージも早いです

noriyotcp commented:
本PRと関係のないコミットは他のPRへと切り出しました。こちらのPRでは

- `Projects#index`, `#search` の作成
- スタイルについて
  - 旧`global_projects` に関するものは `projects` へ
  - 旧`projects#index` に関するものは `owners#show` へ
  - jsも同様に
- `global_projects` に関するコードの全削除
- ~~`.scss-lint.yml`, `.eslintrc.js` の追加と調整~~ #45

noriyotcp commented on app/controllers/projects_controller.rb:
```diff
@@ -1,12 +1,27 @@
 
   authorize_resource
 
+  def index
```
`GlobalProjects#index` からロジックを移動させてきた。`#search` も同様
必要なプライベートメソッドもこちらに移動させた

noriyotcp commented on app/views/layouts/project.html.slim:
```diff
@@ -4,7 +4,7 @@
 - content_for :ogp do
 
-  - if @project.persisted?
+  - if @project&.persisted?
```
`Projects#index`, `#search` では`@project` を返さないので、ぼっち演算子をここで使用している

noriyotcp commented on app/views/layouts/project.html.slim:
```diff
@@ -27,11 +27,13 @@
-  - has_derivatives = @project.derivatives.present?
-  - if has_origin || has_derivatives
-    = render "projects/relations"
+  - if @project&.original.present? || @project&.derivatives.present?
```
`has_origin`, `has_derivatives` の変数は使用しない形に変えた
似たようなロジックが`projects/_relations` パーシャル内にもあったが、そちらも変数の使用をやめた

noriyotcp commented on app/views/layouts/project.html.slim:
```diff
@@ -27,11 +27,13 @@
+    = render partial: "projects/relations", locals: { project: @project }
 
-  #content.project-layout
+  - if current_page?(controller: 'projects', action: 'index') || current_page?(controller: 'projects', action: 'search')
```
プロジェクト一覧、検索結果のレイアウトは、もともとのプロジェクトのレイアウトと微妙に違うので、ここで切り分けている

もともとのレイアウトは`#content.project-layout` 配下に表示される

noriyotcp commented on app/views/owners/show.html.slim:
```diff
@@ -1,6 +1,6 @@
 - content_for(:title, "#{@owner.slug} (#{@owner.name}) : Project list")
 
-#projects-index
+#owners-show
```
旧`Projects#index` が`Owners#show` に移行したのに伴い、スタイルも移動させ、`#owners-show` にリネームした

noriyotcp commented on app/views/projects/_dashboard.html.slim:
```diff
@@ -25,6 +25,8 @@
 
-    - unless @is_searching
-      .label
+    - if current_page?(controller: 'projects', action: 'index')
```
旧`GlobalProjects#index` 内でフラグを使用して切り分けていたのを、`current_page?` を使用する

ちなみに、今までログインした際に日本語設定でも `Featured Projects` のままだったかもしれない・・・

noriyotcp commented on app/views/projects/show.html.slim:
```diff
@@ -2,12 +2,96 @@
-  = render "projects/project_comments"
-  = render "projects/templates"
-  = render "projects/state_convert_dialog"
+  = render partial: "basic_informations"
```
たった１箇所でしかレンダリングされていないパーシャルについてはまとめてしまうことにした
projects/index.html.slim でも同様にまとめている

noriyotcp replied:
⬆️ パーシャルをまとめるのはこのPRがマージされた後にする

noriyotcp commented:
WIP 外しました。主に見てほしい個所は`Projects#index`, `#search` 関連で、コメントをしています

takeyuweb commented:
@noriyotcp 高さのズレが気になるのであわせられますか？
![image](https://user-images.githubusercontent.com/60980/44380977-2f538b00-a549-11e8-94ab-171177afecd2.png)

noriyotcp commented:
@takeyuweb `padding-top: 18px` にしてこのようにしました

![2018-08-21 14 19 28](https://user-images.githubusercontent.com/5820754/44382021-287b4700-a54e-11e8-8510-1ef1f43b714b.png)

takeyuweb reviewed (approved):
内容についてはなんだかよくわからない

noriyotcp commented on app/views/projects/index.html.slim:
```diff
@@ -0,0 +1,219 @@
+        - @projects.each do |project|
+          = render partial: "project", locals: { project: project }
+      .more
+        = link_to "→ More Projects", "/search?page=2", class: "more-projects"
```
ここは旧`global_projects/_reciped_projects.html.slim` でこうなっていたが

https://github.com/takeyuwebinc/gitfab2/blob/4121ab07f7bcf01b30532d4303330a6292bb05e7/app/views/global_projects/_reciped_projects.html.slim#L7-L10

そもそも`@projects` が12個までしか取得されないので、条件に当てはまらず、リンクは表示されていなかった

![2018-08-21 17 24 22](https://user-images.githubusercontent.com/5820754/44389935-1eb20d80-a567-11e8-8b1c-646488017993.png)

クエリなしの検索画面2ページ目へと遷移するリンクにした

noriyotcp commented on app/views/projects/index.html.slim:
```diff
@@ -0,0 +1,219 @@
+        - @projects.each do |project|
+          = render partial: "project", locals: { project: project }
+      .more
+        = link_to "→ More Projects", search_path(page: 2), class: "more-projects"
```
ここは旧`global_projects/_reciped_projects.html.slim` でこうなっていたが

https://github.com/takeyuwebinc/gitfab2/blob/4121ab07f7bcf01b30532d4303330a6292bb05e7/app/views/global_projects/_reciped_projects.html.slim#L7-L10

そもそも`@projects` が12個までしか取得されないので、条件に当てはまらず、リンクは表示されていなかった

![2018-08-21 17 24 22](https://user-images.githubusercontent.com/5820754/44389935-1eb20d80-a567-11e8-8b1c-646488017993.png)

クエリなしの検索画面2ページ目へと遷移するリンクにした

itkrt2y commented:
@noriyotcp `.scss-lint.yml` は「プロジェクト全体で反映されるべきこと」と「このPRでは扱わないこと」をしっかり区別して書いてください。
無効にしたもの、ほどんどはプロジェクトとしては有効であって欲しいものですよね？

itkrt2y commented:
あと、lint系は別PRにしていただいた方がレビューしやすいです。このPRではSiderが上げてくるissueはcloseでいいです。

itkrt2y commented on .scss-lint.yml:
```diff
@@ -155,11 +155,11 @@ linters:
   SelectorFormat:
-    enabled: true
+    enabled: false
     convention: hyphenated_lowercase # or 'classic_BEM', or 'hyphenated_BEM', or 'snake_case', or 'camel_case', or a regex pattern
```
このlintは `convention: snake_case` にするのでいいんじゃないかな

itkrt2y commented on app/assets/stylesheets/mixins/colors.scss:
```diff
@@ -5,11 +5,11 @@ $common-light-gray: #EAEAEA;
 $common-dark-gray: #555555;
 
-$common-text-color: black;
+$common-text-color: #000000;
```
`ColorVariable` のlintの存在理由は、場当たり的に `p { color: red }` とか書くのをやめさせたいってことだと思われますが、それでいうと一度変数に入れて使うのはむしろ好ましいことなので、それをカラーコードに直しちゃったら読みづらくて悪です
https://github.com/brigade/scss-lint/blob/master/lib/scss_lint/linter/README.md#colorvariable

itkrt2y commented:
partialをまとめるのもこのPRの後に別PRとして出してもらえると、PRの責務が分かれていいです

noriyotcp committed:
Add Projects#index and #search

Projects#indexでプロジェクトの一覧を表示
Projects#searchで検索結果を表示

itkrt2y commented:
ヘッダーの位置に関しては修正後もずれてる（そもそももっと根本的に直さなければならないことなので、今はこれでいいですが）
![image](https://user-images.githubusercontent.com/2343568/44501651-212c7880-a6c9-11e8-97ce-792b0f6c1ae5.png)
